feat: improve SourceMapDevToolPlugin

SyMind opened the pull request:
<!--
  Thank you for submitting a pull request!

  We appreciate the time and effort you have invested in making these changes. Please ensure that you provide enough information to allow others to review your pull request.

  Upon submission, your pull request will be automatically assigned with reviewers.

  If you want to learn more about contributing to this project, please visit: https://github.com/web-infra-dev/rspack/blob/main/CONTRIBUTING.md.
-->

## Summary

<!-- Can you explain the reasoning behind implementing this change? What problem or issue does this pull request resolve? -->

<!-- It would be helpful if you could provide any relevant context, such as GitHub issues or related discussions. -->

Fix #5060
Fix #4893

## Test Plan

<!-- Can you please describe how you tested the changes you made to the code? -->

Ensure the normal execution of current source map test cases and migrate more test cases from webpack.

## Require Documentation?

<!-- Does this PR require documentation? -->

- [x] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

SyMind commented:
In this PR, I've attempted to support all options of the `SourceMapDevToolPlugin`, but some options are still not supported so far, such as `fileContext`.

However, due to the extensive changes made, I believe it is now appropriate to proceed with a code review to avoid any potential directional mistakes.

ahabhgk commented on crates/rspack_core/src/options/compiler_options.rs:
```diff
@@ -15,7 +17,7 @@ pub struct CompilerOptions {
   pub resolve_loader: Resolve,
   pub module: ModuleOptions,
-  pub devtool: Devtool,
+  pub devtool: Mutex<Devtool>,
```
The Mutex is not necessary at all, this should be readonly, all options are finalized at js side

SyMind replied:
The `devtool` setting currently is not yet finalized on the JavaScript side. The `SourceMapDevToolPlugin` will configure it completely.

In Webpack, the [SourceMapDevToolModuleOptionsPlugin](https://github.com/webpack/webpack/blob/main/lib/SourceMapDevToolModuleOptionsPlugin.js) is used to set `useSourceMap` or `useSimpleSourceMap` to true, which signals that each module should generate a source map.

In rspack, I've used the `devtool` setting to achieve the aim .

SyMind replied:
It's worth mentioning that a comment in the webpack code suggests the `useSourceMap` and `useSimpleSourceMap`  might be refactored in webpack 6, which is something I'm not very familiar with. You can see the comment here: https://github.com/webpack/webpack/blob/main/lib/Module.js#L169

ahabhgk replied:
it's only a refactor without any logic change, refactor `useSourceMap` and `useSimpleSourceMap` to one enum

SyMind replied:
// TODO refactor this -> options object filled from Factory

ahabhgk replied:
ohh, then we can delay SourceMapDevToolModuleOptionsPlugin, but devtool shouldn't wrapped by Mutex

SyMind replied:
If we don't wrap `devtool` with a Mutex, what method should I use to modify it in `SourceMapDevToolModulePlugin`?

SyMind replied:
@ahabhgk I have used `RwLock` instead of `Mutex`. It's enough for the goal of modifying `devtool` in `compilation` hook and it's more cheap.

SyMind replied:
Do you have any other methods?

ahabhgk commented on crates/rspack_core/src/normal_module.rs:
```diff
@@ -398,6 +402,7 @@ impl Module for NormalModule {
         code_generation_dependencies: &mut code_generation_dependencies,
         build_info: &mut build_info,
         build_meta: &mut build_meta,
+        devtool: &devtool,
```
should have this in ParseResult

SyMind replied:
This `devtool` setting will be used in the parse function, which is not asynchronous.

ahabhgk commented on crates/rspack_plugin_devtool/src/lib.rs:
```diff
@@ -66,88 +200,231 @@ impl Plugin for SourceMapDevToolPlugin {
+      devtool.add_no_sources();
+    }
+    if self.module {
+      devtool.add_module();
```
module should setup by SourceMapDevToolModuleOptionsPlugin

SyMind replied:
I believe it is not the right time to implement `SourceMapDevToolModuleOptionsPlugin`. We should wait for the design to finalize regarding `useSourceMap` and `useSimpleSourceMap` within the module.

ahabhgk commented on crates/rspack_plugin_devtool/src/lib.rs:
```diff
@@ -274,6 +756,24 @@ impl Plugin for EvalSourceMapDevToolPlugin {
+    }
+    if self.no_sources {
+      devtool.add_no_sources();
+    }
```
also need to handle "module" (eval-cheap-module-source-map)

SyMind replied:
The decision might be made in `packages/rspack/src/rspackOptionsApply.ts`.

SyMind commented:
@ahabhgk I've introduced a new trait called `SourceMapConf`. The `Module` trait now extends it. This is intended to implement features similar to webpack's `useSourceMap` and `useSimpleSourceMap`.

Additionally, I've removed the `Devtool` struct as the responsibilities are now handled by the `SourceMapDevToolPlugin` and the `SourceMapConf` trait.

ahabhgk commented on crates/rspack_core/src/options/compiler_options.rs:
```diff
@@ -15,7 +15,7 @@ pub struct CompilerOptions {
   pub resolve_loader: Resolve,
   pub module: ModuleOptions,
-  pub devtool: Devtool,
+  pub devtool: String,
```
This can be removed

ahabhgk commented on crates/rspack_core/src/parser_and_generator.rs:
```diff
@@ -2,6 +2,7 @@ use std::{collections::HashMap, fmt::Debug};
 
 use derivative::Derivative;
 use rkyv::AlignedVec;
+use rspack_common::SourceMapKind;
```
Can we move this into rspack_core?

ahabhgk commented on crates/rspack_plugin_css/src/runtime/mod.rs:
```diff
@@ -10,6 +11,7 @@ use rspack_plugin_runtime::{
 };
 use rustc_hash::FxHashSet as HashSet;
 
+#[impl_runtime_module]
```
We can also improve other code like `DependenciesBlock` by proc macro

ahabhgk reviewed (commented):
Generally looks good to me

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
bench!

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench

SyMind commented:
!bench
